implementation/src/main/java/io/smallrye/jwt/auth/principal/KeyLocationResolver.java

```diff
@@ -128,18 +130,58 @@ private PublicKey tryAsJwk(JsonWebSignature jws) {
     }
 
     PublicKey getHttpsJwk(String kid) {
-        LOGGER.debugf("Trying to create a key from the HTTPS JWK(S)...");
+        try {
+            PublicKey theKey = getKeyFromJsonWebKeys(kid, httpsJwks.getJsonWebKeys(), authContextInfo.getSignatureAlgorithm());
+            if (theKey != null || isMatchingJwkAvailable(httpsJwks.getJsonWebKeys(), kid)) {
```

Comment:
It may be possible for the two calls to `getJsonWebKeys` to return different lists depending on traffic/threads. It might be better to assign a local variable from a single call instead. If `theKey` is null and another thread refreshes the JWKS cache before the `isMatchingJwkAvailable`/`getJsonWebKeys`, I think it may return the null if the `kid` is found in the refreshed list. An edge case, but something to consider.

Reply:
Hi Mike, @MikeEdgar thanks, good catch, I had `List<JsonWebKey> keys = httpsJwks.getJsonWebKeys()` at some point but then I started optimizing :-) and when it is done prematurely, bugs do happen :-)

Reply:
Re the sync, I'd've gladly avoided it but I was explained it is a DOS risk otherwise, someone can keep messing with the kid header and if we do immediately every time it can seriously slow the system

Reply:
But this sync will be rarely hit, only briefly around the rotation period, so it is fine

Reply:
@sberyozkin - that's what I figured as well. I looked over this section several times in conjunction with the implementation of `refresh` (which itself uses a semaphore), and I think the approach you took is good.

Reply:
Ok, thanks, resolving it now, a single call to get the list is made